operator/cmd/certificates/watcher.go

```diff
@@ -0,0 +1,137 @@
+package certificates
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// TODO: refactor code below to be testable and also tested
+const (
+	certificateRenewalInterval = 6 * time.Hour
+	ServerKey                  = "tls.key"
+	ServerCert                 = "tls.crt"
+)
+
+type CertificateWatcher struct {
+	apiReader             client.Reader
+	fs                    afero.Fs
+	certificateDirectory  string
+	namespace             string
+	certificateSecretName string
+	Log                   logr.Logger
+}
+
+func NewCertificateWatcher(mgr manager.Manager, namespace string, secretName string, log logr.Logger) *CertificateWatcher {
+	return &CertificateWatcher{
+		apiReader:             mgr.GetAPIReader(),
+		fs:                    afero.NewOsFs(),
+		certificateDirectory:  mgr.GetWebhookServer().CertDir,
+		namespace:             namespace,
+		certificateSecretName: secretName,
+		Log:                   log,
+	}
+}
+
+func (watcher *CertificateWatcher) watchForCertificatesSecret() {
+	for {
+		<-time.After(certificateRenewalInterval)
+		watcher.Log.Info("checking for new certificates")
+		if updated, err := watcher.updateCertificatesFromSecret(); err != nil {
+			watcher.Log.Info("failed to update certificates", "error", err)
+		} else if updated {
+			watcher.Log.Info("updated certificate successfully")
+		}
+	}
+}
+
+func (watcher *CertificateWatcher) updateCertificatesFromSecret() (bool, error) {
```

Comment:
do we need the bool return value? This seems to only be true if err == nil, so we might simplify this be just returning an error

Reply:
if the bytes are not the same we should still fail somehow, but if you like it better I could rewrite it to return an adhoc error and remove the bool